nixos/grub: Fixes to allow flicker-free boot, themes and bug fixes

samueldr opened the pull request:
###### Description of changes

I can split in further PRs if the changes are judged to warrant further evaluation. The four changes are independent, but were developed when working on flicker-free boot in my personal config. They are not intrinsically linked, but write out a story.

**Always install with bootloader id** fixes an oversight where choosing not to touch EFI variables, but not installing as removable, wouldn't install to the distro-owned path in the ESP. This is problematic if you're including additional components in your boot chain, as it means the binary for GRUB may be installed at a different location depending on an unrelated option. Registering GRUB to the EFI boot order shouldn't change where we install it.

**Fix extraFiles in subdirectories** ensures that files in subdirectories passed to `extraFiles` gets installed. With this, it's possible to add some other component to the ESP, for example `rEFInd` could be copied for problematic systems, or other fun shenanigans like using [HackBGRT](https://github.com/samueldr/HackBGRT) to change the vendor logo used during boot. This change makes it work just like it does with `systemd-boot`.

**Add support for timeoutStyle** allows configuring a GRUB option, making it possible to hide it during boot. Though it looks like `countdown` would be nice, it is not. It will always show a text-mode countdown. Boooh. `hidden` completely hides it, it can be shown with ESC, F4 or by holding shift (as documented upstream). Hidden [with other patches](https://github.com/NixOS/nixpkgs/pull/220261) allows for flicker-free boot.

Finally, **Insert required image modules for themes** ensures that any modules needed for the theme are copied over. Without this, the example config for the `theme` option (`pkgs.nixos-grub2-theme`) wouldn't work.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

samueldr committed:
nixos/grub: Fix extraFiles in subdirectories

This directly copies the systemd-boot logic, which works.

`install` with `-D` will create all leading directory components.

samueldr committed:
nixos/grub: Add support for timeoutStyle

samueldr committed:
nixos/grub: Always install with bootloader id

The intent was to not pass the flag when installing as removable. In
reality there is a third case, where you may not want to touch EFI
variables, and not want to install as removable.

In that case, it would install to the generic \EFI\grub\grubx64.efi,
which is not a good choice in any cases. The operating system should
"own" their path under \EFI\ to be a good citizen [citation needed].

With this change, there can be only two paths GRUB can be installed to:

 - \EFI\NixOS-boot\grubx64.efi
 - \EFI\BOOT\bootx64.efi

This removes the surprising behaviour where GRUB may be installed to a
different location only because we configured NixOS not to touch EFI
variables.

It may be necessary under some configurations to install GRUB without
touching EFI variables, but to the NixOS-owned location.

samueldr committed:
nixos/grub: Insert required image modules for themes

With a limited testing of all packaged GRUB 2 themes (pkgs.nixos-grub2-theme)
this is tested to work.

Without this change, the theme loading will error out (waiting for a key press).

With this change, the theme loads and works as expected.

alyssais commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -57,6 +57,7 @@ let
       storePath = config.boot.loader.grub.storePath;
       bootloaderId = if args.efiBootloaderId == null then "${config.system.nixos.distroName}${efiSysMountPoint'}" else args.efiBootloaderId;
       timeout = if config.boot.loader.timeout == null then -1 else config.boot.loader.timeout;
+      timeoutStyle = cfg.timeoutStyle;
```
Let's reuse the `inherit (cfg)` we already have below instead?

samueldr replied:
https://github.com/NixOS/nixpkgs/pull/220263/files#diff-713d422d999999a630d109add2653608eb7eacade640f61f781ec521c47afb7eR91

samueldr commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -450,6 +451,26 @@ in
+        description = lib.mdDoc ''
+           - `menu` shows the menu.
+           - `countdown` uses a text-mode countdown.
+           - `hidden` hides GRUB entirely.
```
(Note to self) Link to the upstream docs about the hidden mode, in addition to stating which keys can be used to show the menu.

samueldr replied:
See additions under this comment

samueldr committed:
nixos/grub: Remove `with` with broad scopes.

There is only other `with` with a somewhat broad scope, `with pkgs`, but
it's used in a place where it would become awkward to change out. And
anyway its scope is rather limited still.

samueldr commented:
Last push:

 - Rebases on top of current `nixos-unstable`
 - Updates post removal of GRUB 1
 - Addresses issues
 - Removes `with` clauses having too broad of  a scope.

While I hadn't pushed here, I have been daily-driving this for a good while, the rebase post-grub1-removal happened a while back.

So as far as using, it's been well worn.

samueldr commented:
¯\\\_(ツ)\_/¯

nh2 commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -699,7 +747,7 @@ in
       boot.loader.grub.extraPrepareConfig =
         concatStrings (mapAttrsToList (n: v: ''
-          ${pkgs.coreutils}/bin/cp -pf "${v}" "@bootPath@/${n}"
+          ${pkgs.coreutils}/bin/install -Dp "${v}" "${efi.efiSysMountPoint}/"${escapeShellArg n}
```
@samueldr This reads like it will break `mirroredBoots`.

[`boot.loader.efi.efiSysMountPoint`](https://search.nixos.org/options?channel=23.11&show=boot.loader.efi.efiSysMountPoint&from=0&size=50&sort=relevance&type=packages&query=efiSysMountPoint) appears to be a variable that must not be read if `mirroredBoods` is used (because then, there may be multiple [`boot.loader.grub.mirroredBoots.*.efiSysMountPoint`](https://search.nixos.org/options?channel=23.11&show=boot.loader.efi.efiSysMountPoint&from=0&size=50&sort=relevance&type=packages&query=efiSysMountPoint)s that all need to be handled instead).

`efi.efiSysMountPoint` is apparently only read in 1 place: To initialise `boot.loader.grub.mirroredBoots.*.efiSysMountPoint` if `mirroredBoots` are not explicitly defined:

https://github.com/NixOS/nixpkgs/blob/dd37924974b9202f8226ed5d74a252a9785aedf8/nixos/modules/system/boot/loader/grub/grub.nix#L708-L709



But with this change, you're switching to read `efi.efiSysMountPoint` directly, unconditionally.

This seems to directly undo the intention by @Frostman, in the commit that last touched this line: https://github.com/NixOS/nixpkgs/commit/7144b9ac5417c6e74af41fd3c6cad8e20ffc369c

> Fix boot.loader.grub.extraFiles when used with mirroredBoots
>
> Substitute `@bootPath@` in `boot.loader.grub.extraPrepareConfig` script same way as it's done for `boot.loader.grub.extraEntries` option.

Was the `@bootPath@` usage actually wrong, or did you just care about the usage of `-D` and `escapeShellArg`?

If yes, `@bootPath@` should probably be added back.

Also, I think there should be a comment on `efi.efiSysMountPoint` to say that it should not be read, because code written against it will not work with `mirroredBoots`. Where "not work" means "break" if the value is != any of the `mirroredBoots` one, and otherwise mean "silently act on only one of the ESPs, thus removing the benefits of `mirroredBoots` (system may not come back if one of the mirrored devices breaks)".
